src/main/java/com/redhat/labs/lodestar/service/MigrationService.java

```diff
@@ -0,0 +1,151 @@
+    
+    private Map<Integer, Engagement> allEngagements = new HashMap<>();
+    
+    /**
+     * Currently the migration will only occur if config properties are true.
+     * The migration is idempotent so no harm in rerunning. It will only update
+     * engagements that haven't been migrated. As we get closer to migration time
+     * we should evaluate whether this is the right approach.
+     * Once the start up is complete this service can not be called.
+     * @param ev
+     */
+    void onStart(@Observes StartupEvent ev) {
+        
+        if(migrateUsers) {
+            LOGGER.info("Migrate users: {}", migrateUsers);
+            migrateUsers();
+            LOGGER.info("End Migrate users");
+        }
+        
+        if(migrateUuids) {
+            LOGGER.info("Migrate uuids: {}", migrateUuids);
+            migrateUuids();
+            LOGGER.info("End Migrate uuids");
+        }  
+        
+    }
+    /**
+     * Get all projects and split for individual update
+     */
+    private void migrateUuids() {
+        List<Project> allProjects = projectService.getProjectsByGroup(engagementRepositoryId, true);
+        allProjects.parallelStream().forEach(this::updateProjectWithUuid);
+    }
+    
+    /**
+     * Update a single project if the description is not already set and an engagement has been found
+     * @param project
+     */
+    private void updateProjectWithUuid(Project project) {
+        
+        if(project.getDescription() == null && getAllEngagements().containsKey(project.getId())) {
+            String uuid = allEngagements.get(project.getId()).getUuid();
+            project.setDescription(String.format(ProjectStructureService.ENGAGEMENT_PROJECT_DESCRIPTION, uuid));
+            projectService.updateProject(project);
+            
+            LOGGER.info("Added uuid {} to project {} {}", uuid, project.getId(), project);
+        }
+    }
+
+    /**
+     * Get all engagements (engagement.json) and split for individual update
+     */
+    private void migrateUsers() {     
+        getAllEngagements().values().parallelStream().forEach(this::migrateUsersToGitlab);   
+    }
+    
+    /**
+     * Get All engagements. This is run by migrate users and uuids so only run once if both are active
+     * @return
+     */
+    private Map<Integer, Engagement> getAllEngagements() {
+        if (allEngagements.isEmpty()) {
+            List<Engagement> engagements = engagementService.getAllEngagements(Optional.of(false), Optional.of(false));
+            engagements.parallelStream().forEach(this::addToMap);
+            LOGGER.debug("fetched engagements for migration {} ", allEngagements.size());
+        }
+        
+        return allEngagements;
+    }
+    
+    /**
+     * Do a quick project id to engagement mapping to easily find the engagement by project id
+     * @param engagement
+     */
+    private void addToMap(Engagement engagement) {
+        LOGGER.debug("mcanoy {}",engagement);
```

Comment:
not sure if you wanted your id immortalized.

Reply:
whoops